### ISSUE TITLE
Dynamic block render ignores attributes of type object.

### ISSUE THREAD

**Issue by juanRabaa** [WordPress/gutenberg#19889]
**Describe the bug**
Rendering a dynamic block with `render_block` ignores an attribute of type object. It dissappears from the attributes array that the `render_callback` receives

```` php
<?php
//Block registration
register_block_type( 'lr/latest-articles', array( 
	'attributes'      => array( 
		'articles_query'	=> array(
			'type'		=> 'object',
			'default'	=> null,
		),
	), 
	'render_callback' => 'the_render_callback', 
) ); 
````
```` php
<?php
//Block render
echo render_block(array(
    'blockName'            =>  'lr/latest-articles',
    'attrs'                => array(
        'articles_query'         =>  $wp_query //An instance of WP_Query
    ),
    'innerContent'         => [],
));
````
The `articles_query` is suppossed to have an instance of `WP_Query`. If not passed, the `$attributes['articles_query']` is equal to null. If passed, the `$attributes['articles_query']` completely dissappears.

### Problem source ###
Debugging the `render_block` function, it seems that the `$block['attrs']` has the correct code. It changes at `WP_Block_Type::render` method, when line 107 - `$attributes = $this->prepare_attributes_for_render( $attributes )` - is reached.

The problem occurs when validating the attribute, in `WP_Block_Type::prepare_attributes_for_render` `$is_valid = rest_validate_value_from_schema( $value, $schema );` returns `false` for the object attribute with an instance of `WP_Query`.

When validating an attribute of type object, the `rest_validate_value_from_schema` only checks against `stdClass` and `JsonSerializable`. Since the object I'm passing is instanceof `WP_Query`, it fails the validation, and because the default value of the attribute is null, it is remove completely.

```` php
//Portion of the rest_validate_value_from_schema object validation
// $value = instance of WP_Query
if ( 'object' === $args['type'] ) {
	if ( $value instanceof stdClass ) { //false
		$value = (array) $value;
	}

	if ( $value instanceof JsonSerializable ) { //false
		$value = $value->jsonSerialize();
	}

	if ( ! is_array( $value ) ) { // true => fails validation
		/* translators: 1: Parameter, 2: Type name. */
		return new WP_Error( 'rest_invalid_param', sprintf( __( '%1$s is not of type %2$s.' ), $param, 'object' ) );
	}
(...)
````

## Question ##

If this is not a bug, how else am I supposed to pass a `WP_Query` instance (or any other object) to a dynamic block, if not as an attribute of type object?

---------

**To reproduce**
1. Register a dynamic block with an attribute of type object and default `null`
2. Render the block using `render_block` and passing an object to the attribute setted before

**Expected behavior**
The attributes that the `render_callback` receives should be the same sent through `render_block`

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome
 - Version: 79.0.3945.117

**Additional context**
- Please add the version of Gutenberg you are using in the description: The one that comes with Wordpress 5.3.2

**Comment by nhterry** [comment]
@juanRabaa Not sure if this is your issue, but I think you need to specify what type of items are in your object.

```
	'type' => 'object',
		'items' => [
			'type' => 'string',
		],
```

For a nested array / object I used:

```
	'groups'  => array(
		'type' => 'array',
		'default' => array(),
		'items' => [
			'type' => 'object',
			'items' => [
				'type' => 'object',
				'items' => [
					'type' => 'string',
				],
			],
		],
	),
```

**Comment by juanRabaa** [comment]
> @juanRabaa Not sure if this is your issue, but I think you need to specify what type of items are in your object.
> 
> ```
> 	'type' => 'object',
> 		'items' => [
> 			'type' => 'string',
> 		],
> ```
> 
> For a nested array / object I used:
> 
> ```
> 	'groups'  => array(
> 		'type' => 'array',
> 		'default' => array(),
> 		'items' => [
> 			'type' => 'object',
> 			'items' => [
> 				'type' => 'object',
> 				'items' => [
> 					'type' => 'string',
> 				],
> 			],
> 		],
> 	),
> ```

I don't think this is the problem. Even if it were, I wouldn't know how am I supposed to specify the object "items" type if it has mixed properties, like a WP_Query instance (and most objects) do.

**Comment by talldan** [comment]
Hi @juanRabaa. The block attributes need to be serialisable as json as they're stored in the post content like this (an example with the latest posts block):
![Screen Shot 2020-02-06 at 2 57 11 pm](https://user-images.githubusercontent.com/677833/73913078-14b5d300-48f1-11ea-8342-0b3eb68ebb92.png)

If WP_Query isn't serialisable you may need to convert it into an object that can be serialised.

I think this could be explained better as part of the documentation (perhaps here: https://developer.wordpress.org/block-editor/tutorials/block-tutorial/creating-dynamic-blocks/), so I'll leave this issue open as a documentation task.

**Comment by youknowriad** [comment]
Closing as answered: block attributes have to be JSON-serializable because they live in post content, so a WP_Query instance can't be stored directly. The Block Attributes reference now covers this. Thanks!
